arc

```diff
@@ -0,0 +1,163 @@
+#!/bin/bash
+
+# set -e
+
+usage () {
+  echo "Usage: $0 [OPTIONS] [check|build|run] <INPUT-FILE> [-- <MLIR-OPTS>]"
+  echo "Options:"
+  echo "  -h       Show this help message"
+  echo "  -o <DIR> Output directory (defaults to ./target/)"
+  exit 0
+}
+
+if [ "$DEBUG" ]; then
+  debug () {
+    echo "$@"
+  }
+else
+  debug () {
+    :
+  }
+fi
+
+while getopts ho: opt; do
+    case $opt in
+        h)
+            usage
+            ;;
+        o)
+            CRATE_DIR="$OPTARG"
+            ;;
+        \?)
+            echo "Invalid option: $OPTARG" >&2
+            usage
+            exit 1
+            ;;
+    esac
+done
+
+shift "$((OPTIND - 1))"
+
+debug "Parsed options"
+
+while [[ $# -gt 0 ]]; do
+    case $1 in
+        --)
+            shift
+            break
+            ;;
+        check)
+            MODE=check
+            INPUT_FILE="$1"
+            shift 3
+            break
+            ;;
+        build)
+            MODE=build
+            INPUT_FILE="$2"
+            shift 3
+            break
+            ;;
+        run)
+            MODE=run
+            INPUT_FILE="$2"
+            shift 3
+            break
+            ;;
+        *)
+            echo "Invalid subcommand: $1" >&2
+            usage
+            exit 1
+            ;;
+    esac
+done
+
+debug "Parsed subcommand"
+
+CRATE_NAME="$(basename $INPUT_FILE .arc)-crate"
+
+if [ -z "$INPUT_FILE" ]; then
+    usage
+fi
+
+if [ -z "$CRATE_DIR" ]; then
+    CRATE_DIR="$(dirname $INPUT_FILE)/$CRATE_NAME"
+fi
+
+ARC_DIR=$(realpath $(dirname "$(which "$0")")/)
+
```

Comment:
```shell
mkdir /tmp/a /tmp/b
cat >> /tmp/a/foo <<EOF
#!/bin/bash
echo "Realpath says: $(realpath $(dirname "$(which "$0")"))"
echo "Readlink says: $(dirname "$(readlink -f "$(which "$0")")")"
EOF
chmod 755 /tmp/a/foo
ln -s /tmp/a/foo /tmp/b/foo
```
Then try to run `/tmp/b/foo`:
```shell
$ /tmp/b/foo 
Realpath says: /tmp/b
Readlink says: /tmp/a
```
and you'll see that realpath gets it wrong. Check [how it can be done in a way that is known to work both on Linux, MacOs and even Wintendo](https://github.com/frej/fast-export/blob/dbb8158527719e67874e0bba37970e48e1aaedfb/hg-fast-export.sh#L6).